pypuppetdb/api/pql.py

```diff
@@ -0,0 +1,152 @@
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+import logging
+import re
+from datetime import datetime
+
+import pypuppetdb
+from pypuppetdb.QueryBuilder import (EqualsOperator)
+from pypuppetdb.api.base import BaseAPI
+from pypuppetdb.errors import (APIError)
+from pypuppetdb.types import (Catalog, Node, Report)
+
+log = logging.getLogger(__name__)
+
+
+class PqlAPI(BaseAPI):
+
+    def _pql(self, pql, request_method='GET'):
+        """This method prepares a PQL query to PuppetDB. Actual making
+        the HTTP request is done by _make_request().
+
+        :param pql: PQL query
+        :type pql: :obj:`string`
+
+        :param request_method: (optional) GET or POST, the default is GET
+
+        :raises: :class:`~pypuppetdb.errors.EmptyResponseError`
+
+        :returns: The decoded response from PuppetDB
+        :rtype: :obj:`dict` or :obj:`list`
+        """
+
+        log.debug(f"_pql called with pql={pql}, request_method={request_method}")
+
+        pql = pql.strip()
+        if not pql:
+            log.error("Non-empty PQL query is required!")
+            raise APIError
+
+        payload = {}
+
+        # PQL queries are made to the same endpoint regardless of the queried entities
+        url = self._url('pql')
+        payload['query'] = pql
+
+        return self._make_request(url, request_method, payload)
+
+    # TODO: deduplicate this - see QueryAPI.nodes()
+    def pql(self, pql, with_status=False, unreported=2, with_event_numbers=True):
```

Comment:
don't feel right to have node specific parameters for such a generic pql method.

Reply:
Same here, but I don't have a better idea..